### ISSUE TITLE
 New Feature: Differential Evolu, saving the state-population

### ISSUE THREAD

**Issue by arita37** [scipy/scipy#6517]
Hi,

Is it possible to add a feature to save the state of differential evolution after a number of iteration is done.
http://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.differential_evolution.html

and reload the state from file. It would allow to do batch optimization for very large training.
I think this is not too difficult to implement.

Thanks, regards

**Comment by andyfaff** [comment]
I'm not sure how useful a feature this would be.  Firstly, the `DifferentialEvolutionSolver` object could be pickled/unpickled to do this.  Secondly, the `population_energies` and `population` attributes could be saved to file, then reloaded and the attributes set again.
However, with the latter option the population energies and population (and all the other attributes)  only make sense for the specific minimisation that the `DifferentialEvolutionSolver` was created for.
Could you clarify what you mean by
> batch optimization for very large training

?

**Comment by arita37** [comment]
Hi,

I have an optimization which can last several days of computing
because ojective function is complicated/slow.

I just want to stop the optimization after some steps, save it
and re-start later from that state (because the machine needs to be used by somebody else).

It also helps if I am doing paralell island to share with other instance of DE.

Thanks
Kevin

**Comment by andyfaff** [comment]
>
> I have an optimization which can last several days of computing
> because ojective function is complicated/slow.
>
> I just want to stop the optimization after some steps, save it
> and re-start later from that state (because the machine needs to be used
> by somebody else).
>

In that circumstance I would create a DifferentialEvolutionSolver. In later
versions this solver is an iterator. Here would be some pseudocode to save
state every iteration:

    solver = DifferentialEvolutionSolver(func, bounds)
    for i in range(100):
        best_x, best_cost = next(solver)
        pickle.dump(solver, 'solver_%d.pkl' % i)

When you want to restart the solver you just unpickle the object and start
iterating again.


> It also helps if I am doing paralell island to share with other instance
> of DE.
>

Do you mean that you would like different subpopulations working on the
same problem? Do those subpopulations calculate in parallel and synchronise
somehow after each iteration? There are papers that do this kind of
[parallelisation](
http://www.math.upatras.gr/~npav/papers/TasoulisPPV2004.pdf). There are
several approaches to achieve parallelisation of calculation. However,
parallelisation does not necessarily lead to improved convergence
efficiency. I suppose that you could use two solvers working in parallel to
do something along these line at the moment. All you need to do is select
the members you want to swap between populations and move them.

**Comment by arita37** [comment]
Hello Andrew,

Thanks for your insight.
Is next can be applied to solver as shown below ?

Yes, different population and swap them.
As this is multi-thread, it just increases the number of population...

Thanks, regards

**Comment by andyfaff** [comment]
next (DifferentialEvolutionSolver) is available in 0.18 onwards.

**Comment by andyfaff** [comment]
@arita37, the following gist performs parallel differential evolution over subpopulations. Is this the kind of thing you are using?
@dlax, do you think it would be worth adding this to the `optimize.differential_evolution` function? The default could be `workers=1`.
https://gist.github.com/andyfaff/9cf87fa90c394b7ed02a5cd3432a6189

**Comment by andyfaff** [comment]
With 4 workers and with a 0.01s sleep put into the `rosen` function, the speedup is a factor of 1.4 compared to the normal differential evolution code on 1 processor.

**Comment by arita37** [comment]
Hi,

This might be refine, this is the point of doing multi-processing.
Just wondering why time sleep is necessary here ?

Thanks



On 27 Aug 2016, at 12:13, Andrew Nelson <notifications@github.com> wrote:

With 4 workers and with a 0.01s sleep put into the rosen function, the speedup is a factor of 1.4 compared to the normal differential evolution code on 1 processor.

―

**Comment by andyfaff** [comment]
I used `sleep` to simulate a complicated objective function.
Multiprocessing adds a lot of overhead which is only recovered if the
objective function takes a long time to calculate.

**Comment by geofiber** [comment]
How to solve "pool objects cannot be passed between processes or pickled" issue?
https://stackoverflow.com/questions/59989188/scipy-optimize-multiprocessing-and-resuming-on-last-iteration-in-python

**Comment by andyfaff** [comment]
@geofiber I posted an answer to the stackoverflow question. Bear in mind that `DifferentialEvolutionSolver` is not part of the scipy public API and may be subject to change.
